### PR TITLE
Fix potential deadlocking usage of subprocess.Popen

### PR DESCRIPTION
**Pull request by ferringb**
Documented, albeit subtle, gotcha of Popen objects is that you should
not feed input to the process if it's possible for the process to get
blocked waiting on stdout/stderr writing.  On linux this is only possible
if stdout/stderr wrote something like more than 64k without finishing consuming
stdin.

Either way, it's an obvious fix w/out any risks; thus sorting it.

See http://docs.python.org/2/library/subprocess.html#subprocess.Popen.kill ; the details are in the doc descriptions for stdin/stdout/stderr (that follow that anchor, specifically).

### CODE REVIEW

**Comment by mpdehaan** [comment]
Can you please show the scenario in which case you encountered a lock here so we can reproduce the problem you encountered?

Thanks!

**Comment by ferringb** [comment]
If you require it, I can write a unittest to demonstrate the flaw; that's not going to happen for a week or two since it's not exactly high priority for the work I'm doing.  Just to be clear, the subprocess API misuse is demonstrated via thus:

```
import subprocess
p = subprocess.Popen(
    "for x in `seq 1 10000`; do echo I am the very model of a modern major general; done; read",
    shell=True, stdin=subprocess.PIPE, stdout=subprocess.PIPE)
print "process started; writing.  Note it blocks."
p.stdin.write("x" * 100000)
print "child has the input"
print "consuming"
stdout, stderr = p.communicate()
print len(stdout)
```

python process forks a child, the child is single thread and non async, thus it's either reading stdin, or writing stdout. Whether that output be a traceback, or what, it doesn't matter.  Point is, the parent is blocked waiting to write data to stdin of the child, the child is blocked waiting to finish writing to it's stdout- pipe buffers aren't infinite, thus this occurs and is a literal deadlock.

Basically, this is one of those gotchas w/ subprocess that once you've been bit by it you spot it in other codebases; the docs address it, but it's an easier mistake if you've not been bit by it- after you have been bit, it's quick to spot in the code (which is the sole reason I spotted it/split this patch- I've not triggered the scenario, I just know its there).  The snipset above should make it obvious- hopefully obvious enough I don't have to go digging into the runner guts to try and generate a unit tests.

If you still want a unittests, if someone can write it quicker than I that would be appreciated; I do not think this is worthy of a unit test- especially since I've not got the time/priority to do it right now- but who knows, someone else may be willing to do that work.

**Comment by jimi-c** [comment]
This seems more academic then a real concern, as we have had zero reports nor do I see any situations where we feed more than 64k to a process like this. One other minor thing, your patch results in incorrect padding spaces (only 2 spaces are used), so I'm going to close this for now. If someone actually hits this bug, we can revisit it.
